bundles/org.openhab.binding.automower/src/main/java/org/openhab/binding/automower/internal/bridge/AutomowerBridgeHandler.java

```diff
@@ -0,0 +1,160 @@
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_BRIDGE);
+
+    private @Nullable ScheduledFuture<?> automowerBridgePollingJob;
+    private long automowerPollingIntervalS = TimeUnit.HOURS.toSeconds(2);
+
+    private @NonNullByDefault({}) AutomowerBridge bridge;
+    private @NonNullByDefault({}) AutomowerBridgeConfiguration bridgeConfiguration;
+
+    private final HttpClient httpClient;
+
+    @NonNullByDefault
+    private static class AutomowerBridgePollingRunnable implements Runnable {
+        private final Logger logger = LoggerFactory.getLogger(AutomowerBridgePollingRunnable.class);
+
+        private final AutomowerBridgeHandler handler;
+        private final AutomowerBridge bridge;
+
+        private AutomowerBridgePollingRunnable(AutomowerBridgeHandler handler, AutomowerBridge bridge) {
+            this.handler = handler;
+            this.bridge = bridge;
+        }
+
+        @Override
+        public void run() {
+            MowerListResult automowers;
+            try {
+                automowers = bridge.getAutomowers();
+                handler.updateStatus(ThingStatus.ONLINE);
+                logger.info("Found {} automowers", automowers.getData().size());
+            } catch (AutomowerCommunicationException e) {
+                handler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "@text/comm-error-query-mowers-failed");
+                logger.warn("Unable to fetch automowers", e);
+            }
+        }
+    }
+
+    public AutomowerBridgeHandler(Bridge bridge, HttpClient httpClient) {
+        super(bridge);
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Handler disposed.");
+        if (bridge != null) {
+            stopAutomowerBridgePolling(bridge);
+            bridge = null;
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing automower bridge handler.");
+        bridgeConfiguration = getConfigAs(AutomowerBridgeConfiguration.class);
+
+        final String appKey = bridgeConfiguration.getAppKey();
+        final String userName = bridgeConfiguration.getUserName();
+        final String password = bridgeConfiguration.getPassword();
+        if (appKey == null || appKey.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "@text/conf-error-no-app-key");
+        } else if (userName == null || userName.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "@text/conf-error-no-username");
+        } else if (password == null || password.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "@text/conf-error-no-password");
+        } else {
+            updateStatus(ThingStatus.UNKNOWN);
```

Comment:
Is this enough to move it to the end of the method or should the handler methods be synchronized?
Because when multiple threads interact with the object and members might be updated from different threads I guess we would need some synchronization to ensure memory visibility accross threads?

Reply:
It's enough to move it to the end. The framework takes care of thread-safety until setting it to unknown. The handlers shouldn't be synchronized but the code within should be thread-safe. I.e. using synchronized Lists, ConcurrentHashMap etc. if they can be updated concurrently.

Reply:
ok